### PR TITLE
Add emoji reactions and reporting for messages (website)

### PR DESCRIPTION
**Pull request by othrayte**
Implements web side of #917

Move labeling to message ... menu and add reporting and emoji reactions

This new menu appears to the right of the message, and floats down if it doesn't fit. The emoji reactions appear to the left of the menu button and float down with the button when there isn't enough space. The menu contains a list of emoji the user can react with (currently just thumbs up and down) formatted as a quick access grid.

The user can also click on an existing emoji to remove them from it or add them to its count, the emoji disappears when the count reaches 0.

Also added "open in new tab" option so that messages can be navigated to from tasks on mobile.

emoji1
emoji2

Notes:
 * There is a bug where the checked status for emojis is missing from the recent messages list. This appears to be due to a limitation of the backend were we need to provide a user to get emoji for, but doing so filters the list to only contain messages from that user.
 * We can add more emoji easily in future, we just need to pick ones that we have consistent icons for.
 * A future task will add some buttons to the report popup to allow the user to select labels to apply with the report (i.e. spam, contains PII etc)
 * The backend adds a flag emoji when the user files a report, the user can remove their flag emoji, but I don't think the backend does anything with it. The user doesn't currently see the flag emoji until they refresh or change another emoji (this is fixable but there are higher priority things I need to work on first),
 * Other users can currently click on other users flag emoji to also flag, but at the moment they don't get a popup to fill anything in.
 * Currently the emojis are updated when they are changed, this greatly simplifies the UI code but means that the user can technically see other people adding reactions when they do, as rare as that probably is.

### CODE REVIEW

File: website/src/components/Messages/LabelPopup.tsx [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      <ModalOverlay />
+      <ModalContent>
+        <ModalHeader>Label</ModalHeader>
+        <ModalCloseButton />
+        <ModalBody>
+          <LabelInputGroup labelIDs={valid_labels.map(({ name }) => name)} onChange={setValues} />
+        </ModalBody>
+        <ModalFooter>
+          <Button colorScheme="blue" mr={3} onClick={submit}>
+            Submit
```
Comment (fozziethebeat) [inline website/src/components/Messages/LabelPopup.tsx:68]: Any chance you can use the i18n stuff here?

File: website/src/components/Messages/MessageEmojiButton.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import React from "react";
```
Comment (fozziethebeat) [inline website/src/components/Messages/MessageEmojiButton.stories.tsx:1]: Yay stories~

File: website/src/pages/api/report.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Adds a report for a message
+ *
+ */
+const handler = withoutRole("banned", async (req, res, token) => {
+  // Parse out the local message_id, and the interaction contents.
+  const { message_id, text } = req.body;
+
+  const interactionRes = await fetch(`${process.env.FASTAPI_URL}/api/v1/text_labels`, {
```
Comment (fozziethebeat) [inline website/src/pages/api/report.ts:11]: Can this go in the oasst library?

**Review by fozziethebeat — CHANGES_REQUESTED** [review]
Nice! This is a lot of changes but it was pretty easy to review.  Please at least comment on the few comments I made and I'd like to wait for @AbdBarho to review since this is a lot.

File: website/src/components/Messages/LabelPopup.tsx (hunk 1 above)
Comment (othrayte) [reply website/src/components/Messages/LabelPopup.tsx:68]: Sure, I'll take a pass over it and see if I can find them all.

File: website/src/types/Conversation.ts [hunk 4]
```diff
@@ -1,4 +1,16 @@
...
+  name: string;
+  count: number;
+}
+
+export interface MessageEmojis {
+  emojis: { [emoji: string]: number };
+  user_emojis: string[];
+}
+
+export interface Message extends MessageEmojis {
```
Comment (AbdBarho) [inline website/src/types/Conversation.ts:13]: the `Message` interface is what gets received from the backend, does the backend also send the emojis now?

if not, please create a separate types that combines message & emojis.

**Review by AbdBarho — COMMENTED** [review]
Thank you, this was definitely a lot of work!!

File: website/src/pages/api/report.ts (hunk 3 above)
Comment (othrayte) [reply website/src/pages/api/report.ts]: Done

File: website/src/components/Messages/LabelPopup.tsx (hunk 1 above)
Comment (othrayte) [reply website/src/components/Messages/LabelPopup.tsx]: Think I got them all.

File: website/src/components/Messages/MessageEmojiButton.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import { Button } from "@chakra-ui/react";
+import { BoxSelect, Flag, LucideProps, ThumbsDown, ThumbsUp } from "lucide-react";
+import { ReactElement } from "react";
+import { MessageEmoji } from "src/types/Conversation";
+
+type EmojiIconPurpose = "MINI_BUTTON" | "NORMAL";
```
[the remark this answers is not in the document]
Comment (othrayte) [reply website/src/components/Messages/MessageEmojiButton.tsx:6]: TBH that might depend on whether we need to add extra styling, at the moment it just changes the size, but the differentiation is between putting it in normal parts of the UI and putting it in the little buttons on the message. I envisioned that we might have other changes like color and stuff but haven't needed it yet.

**Comment by andreaskoepf** [comment]
> * The backend adds a flag emoji when the user files a report, the user can remove their flag emoji, but I don't think the backend does anything with it. The user doesn't currently see the flag emoji until they refresh or change another emoji (this is fixable but there are higher priority things I need to work on first),

We could for example disallow directly setting the red-flag for ordinary users, so that it only would be set through reporting the message... users could then not un-set the flag on their own.

File: website/src/components/Messages/MessageEmojiButton.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      onClick={onClick}
+      variant={checked ? "solid" : "ghost"}
+      colorScheme={checked ? "blue" : undefined}
+      size="sm"
+      height="1.6em"
+      minWidth={0}
+      padding="0"
+    >
+      {getEmojiIcon(emoji.name, "MINI_BUTTON")}
+      <span style={{ position: "relative", left: "-0.2em", marginRight: "0.25em" }}>{emoji.count}</span>
```
[the remark this answers is not in the document]
Comment (othrayte) [reply website/src/components/Messages/MessageEmojiButton.tsx:45]: Ok, those are new to me, I'll take a look.

**Review by AbdBarho — APPROVED** [review]

File: website/src/types/Conversation.ts (hunk 4 above)
Comment (othrayte) [reply website/src/types/Conversation.ts:13]: Yep, we get emojis for all messages.

**Comment by othrayte** [comment]
@AbdBarho, thanks so much for fixing those last issues 🙏. I had started on it but had to stop and have an early night as I had a splitting headache and nausea by that point.

**Pull request merged**